selendroid-server/src/main/java/io/selendroid/server/model/AndroidNativeElement.java

```diff
@@ -254,18 +256,73 @@ public String getText() {
   public void click() {
     waitUntilIsDisplayed();
     scrollIntoScreenIfNeeded();
+
+    final View view = getView();
+
+    final int[] xy = new int[2];
+
+    if (Build.VERSION.SDK_INT < 19 || view.isLaidOut()) {
+      if (Build.VERSION.SDK_INT < 19) {
+        try {
+          // This is needed for recalculation of location when we cannot
+          // explicitly check if the view's been laid out
+          Thread.sleep(300);
+        } catch (InterruptedException e) {}
+      }
+
+      view.getLocationOnScreen(xy);
+      clickOnScreen(xy[0] + view.getWidth() / 2.0f, xy[1] + view.getHeight() / 2.0f);
+      return;
+    }
+
+    final AndroidWait wait = instrumentation.getAndroidWait();
+    final BooleanWrapper isLaidOut = new BooleanWrapper();
+    final ViewTreeObserver observer = view.getViewTreeObserver();
+    observer.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
+      @Override
+      public void onGlobalLayout() {
+        try {
+          isLaidOut.value = true;
+        } finally {
+          if (observer.isAlive()) {
+            removeOnGlobalLayoutListener(observer, this);
+          } else {
+            removeOnGlobalLayoutListener(view.getViewTreeObserver(), this);
+          }
+        }
+      }
+    }); 
     try {
-      // is needed for recalculation of location
-      Thread.sleep(300);
-    } catch (InterruptedException e) {}
-    int[] xy = new int[2];
-    getView().getLocationOnScreen(xy);
-    final int viewWidth = getView().getWidth();
-    final int viewHeight = getView().getHeight();
-    final float x = xy[0] + (viewWidth / 2.0f);
-    float y = xy[1] + (viewHeight / 2.0f);
+      wait.until(new Function<Void, Boolean>() {
+        @Override
+        public Boolean apply(Void input) {
+          return isLaidOut.value;
+        }
+      });
+    } catch (TimeoutException e) {
+      throw new SelendroidException("View was never laid out", e);
```

Comment:
I'm not quite sure if there are some standard ways of dealing with this, but it would seem useful to include some information about the view that this failed for. I guess this may also be handled on a higher level.